tests/federation_room_event_auth_test.go

```diff
@@ -0,0 +1,206 @@
+package tests
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/federation"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+func TestInboundFederationRejectsEventsWithRejectedAuthEvents(t *testing.T) {
+	/* These tests check that events which refer to rejected events in auth_events
+	 * are themselves rejected.
+	 *
+	 * In order to inject an outlier, we include it as an extra auth_event in a
+	 * regular event. Doing so means that the regular event should itself be
+	 * rejected.
+	 *
+	 * We finish up by sending a final, normal, event which should be accepted
+	 * everywhere. This acts as a sentinel so that we can be sure that the
+	 * events have all been correctly propagated.
+	 *
+	 * The DAG ends up looking like this:
+	 *
+	 *       C
+	 *     / | \
+	 *    /  R  \
+	 *   |   ^   \
+	 *   |    .. O
+	 *   |       ^
+	 *   X .......
+	 *   |
+	 *   S
+	 *
+	 * Where:
+	 *    .... represents an "auth_event" link
```

Comment:
What do the solid lines (`|` `/` `\`) denote? Guessing `prev_events`.

Not sure what the distinction between `prev_events` and `auth_events` is. "This is the last thing I saw" versus "This is what enables me to send this event"?

Reply:
yes, exactly, on both fronts.

auth events are defined by the spec here: https://spec.matrix.org/unstable/server-server-api/#auth-events-selection.

Reply:
added a comment to define `|`.